src/loader/bgw_message_queue.c

```diff
@@ -0,0 +1,373 @@
+	return message_sent;
+}
+
+static BgwMessage * queue_remove(MessageQueue * queue)
+{
+	BgwMessage *message = NULL;
+
+	LWLockAcquire(queue->lock, LW_EXCLUSIVE);
+	if (queue_get_reader(queue) != MyProcPid)
+		ereport(ERROR, (errmsg("cannot read if not reader for TimescaleDB background worker message queue")));
+
+	if (queue->num_elements > 0)
+	{
+		message = palloc(sizeof(BgwMessage));
+		memcpy(message, &queue->buffer[queue->read_upto], sizeof(BgwMessage));
+		queue->read_upto = (queue->read_upto + 1) % BGW_MQ_MAX_MESSAGES;
+		queue->num_elements--;
+	}
+	LWLockRelease(queue->lock);
+	return message;
+}
+
+/*Construct a message*/
+static BgwMessage * bgw_message_create(BgwMessageType message_type, Oid db_oid)
+{
+	BgwMessage *message = palloc(sizeof(BgwMessage));
+	dsm_segment *seg;
+
+	seg = dsm_create(BGW_ACK_QUEUE_SIZE, 0);
+
+	*message = (BgwMessage)
+	{
+		.message_type = message_type,
+			.sender_pid = MyProcPid,
+			.db_oid = db_oid,
+			.ack_dsm_handle = dsm_segment_handle(seg)
+	};
+
+	return message;
+}
+
+
+/* Our own version of shm_mq_wait_for_attach that waits with a timeout so that should our counterparty die before attaching,
+ * we don't end up hanging.*/
+static shm_mq_result
+ts_shm_mq_wait_for_attach(MessageQueue * queue, shm_mq_handle *ack_queue_handle)
+{
+	int			n;
+	PGPROC	   *reader_proc;
+
+
+	for (n = 1; n <= BGW_MQ_NUM_WAITS; n++)
+	{
+		/* The reader is the sender on the ack queue */
+		reader_proc = shm_mq_get_sender(shm_mq_get_queue(ack_queue_handle));
+		if (reader_proc != NULL)
+			return SHM_MQ_SUCCESS;
+		else if (queue_get_reader(queue) == InvalidPid)
+			return SHM_MQ_DETACHED; /* Reader died after we enqueued our
+									 * message */
+#if PG96
+		WaitLatch(MyLatch, WL_LATCH_SET, BGW_MQ_WAIT_INTERVAL);
+#else
+		WaitLatch(MyLatch, WL_LATCH_SET, BGW_MQ_WAIT_INTERVAL, WAIT_EVENT_MQ_INTERNAL);
+#endif
+
+		ResetLatch(MyLatch);
+		CHECK_FOR_INTERRUPTS();
+	}
+	return SHM_MQ_DETACHED;
+
+
+}
+static bool
+enqueue_message_wait_for_ack(MessageQueue * queue, BgwMessage * message, shm_mq_handle *ack_queue_handle)
+{
+	Size		bytes_received = 0;
+	QueueResponseType send_result;
+	bool	   *data = NULL;
+	shm_mq_result mq_res;
+	bool		ack_received = false;
+
+
+	/*
+	 * We don't want the process restarting workers to really distinguish the
+	 * reasons workers might or might not be restarted, and we don't really
+	 * want them to error when workers can't be started, as there are multiple
+	 * valid reasons for that. So we'll simply return false for the ack even
+	 * if we can't attach to the queue etc.
+	 */
+	send_result = queue_add(queue, message);
+	if (send_result != MESSAGE_SENT)
+		return false;
+
+	mq_res = ts_shm_mq_wait_for_attach(queue, ack_queue_handle);
+	if (mq_res != SHM_MQ_SUCCESS)
+		return false;
+
+	/*
+	 * We can use a blocking wait now because if our counterparty detaches we
+	 * will wake
+	 */
+	mq_res = shm_mq_receive(ack_queue_handle, &bytes_received, (void **) &data, false);
```

Comment:
What about a deadlock or something in the other process? I.e., it doesn't detach, just hangs.